### PR TITLE
Implement std::error::Error for tower-h2 error types

### LINKED ISSUE #15: Errors don't implement Error (nor Fail)
I guess this is simply due to early development state, but still.

Working with the returned errors is somewhat uncomfortable, as they don't implement `std::error::Error`, nor anything similar (like `failure::Fail`). Whenever I try to integrate them into something designed to handle errors, it refuses to compile.

### PR DESCRIPTION
**Pull request by hawkw**
Closes #15. Also likely a prerequisite for https://github.com/tower-rs/tower-grpc/issues/6#issuecomment-355347465.

### CODE REVIEW

**Comment by vorner** [comment]
Just a tiny nitpick ‒ some of the descriptions and `Display`s seem to start with lower-case and some with upper-case letter.

**Comment by hawkw** [comment]
Ah, yeah, thanks for catching that @vorner!

File: src/server/mod.rs [hunk 1]
```diff
@@ -8,6 +8,10 @@ use http::{self, Request, Response};
 use tokio_io::{AsyncRead, AsyncWrite};
 use tower::{NewService, Service};
 
+use std::fmt;
+// XXX: alternatively, we could rename the `Error` struct in this module to 
+//      avoid renaming this import.
+use std::error::Error as StdError;
```
Comment (carllerche) [inline src/server/mod.rs]: This is fine... or not bothering to import it at all... either way.

**Comment by hawkw** [comment]
@vorner, 433786a6c99b9e46269568536e9df601d6d8a882 makes case more consistent across `fmt::Display` and `Error::description()` impls; I've tried to follow the standard library's example and make `description()` strings begin with lower-case letters, and the `Display` formatter produce strings with upper-case letters, even when this makes the code less DRY.

**Pull request merged**